File: rust/src/core/audit.rs
use crate::config::EnvdoctorConfig;
use crate::detectors::{Detector, IndexedModel};
use crate::models::{Finding, Origin, ProjectModel, Severity};
use crate::utils::glob::{matches_any_glob, matches_glob};
use std::collections::HashMap;

/// Detector ids and their associated rules in a well-known order. The order
/// here is the order findings are produced and (roughly) the priority.
pub fn all_detectors() -> Vec<Box<dyn Detector>> {
    vec![
        Box::new(crate::detectors::MissingDetector),
        Box::new(crate::detectors::UnusedDetector),
        Box::new(crate::detectors::UndefinedSourceDetector),
        Box::new(crate::detectors::DuplicatesDetector),
        Box::new(crate::detectors::EnvironmentDiffDetector),
        Box::new(crate::detectors::TypeMismatchDetector),
        Box::new(crate::detectors::PublicPrefixDetector),
        Box::new(crate::detectors::WeakSecretDetector),
        Box::new(crate::detectors::TypoDetector),
        Box::new(crate::detectors::SchemaValidationDetector),
    ]
}

/// Run the full audit pipeline over a model and return aggregated findings.
pub fn run_audit(
    _model: &ProjectModel,
    config: &EnvdoctorConfig,
    index: &IndexedModel,
) -> Vec<Finding> {
    let mut findings: Vec<Finding> = Vec::new();
    for detector in all_detectors() {
        findings.extend(detector.detect(index));
    }

    // Apply per-rule severity overrides first.
    findings = apply_rule_severities(findings, config);

    // Remove findings for explicitly ignored variable names.
    findings = apply_ignores(findings, config);

    findings
}

/// Apply user-configured severity overrides for each rule. "off" removes the
/// finding entirely, "error"/"warning" rewrite its severity.
fn apply_rule_severities(findings: Vec<Finding>, config: &EnvdoctorConfig) -> Vec<Finding> {
    findings
        .into_iter()
        .filter_map(|f| {
            let override_sev = config.rules.get(&f.rule_id);
            match override_sev {
                Some(crate::config::RuleSeverity::Off) => None,
                Some(crate::config::RuleSeverity::Error) => {
                    Some(Finding { severity: Severity::Error, ..f })
                }
                Some(crate::config::RuleSeverity::Warning) => {
                    Some(Finding { severity: Severity::Warning, ..f })
                }
                None => Some(f),
            }
        })
        .collect()
}

/// Drop findings whose variable matches any `ignore_variables` glob.
fn apply_ignores(findings: Vec<Finding>, config: &EnvdoctorConfig) -> Vec<Finding> {
    if config.ignore_variables.is_empty() {
        return findings;
    }
    findings
        .into_iter()
        .filter(|f| !matches_any_glob(&config.ignore_variables, &f.variable))
        .collect()
}

/// Helper: is `name` defined in any dotenv file?
pub fn is_defined_in_any_env(index: &IndexedModel, name: &str) -> bool {
    index.env_definitions.contains_key(name)
}

/// Helper: is `name` documented in `.env.example`?
pub fn is_documented(index: &IndexedModel, name: &str) -> bool {
    index.example_names.contains(name)
}

/// Build the `rule_id -> override severity` map for convenient lookups.
#[allow(dead_code)]
pub fn rule_severity_map(config: &EnvdoctorConfig) -> HashMap<String, crate::config::RuleSeverity> {
    config.rules.clone()
}

/// Test helper to check whether a name matches a single ignore glob.
#[allow(dead_code)]
pub fn is_ignored(name: &str, patterns: &[String]) -> bool {
    patterns.iter().any(|p| matches_glob(name, p))
}

/// Flatten all origins for a name from a model (used by detectors needing
/// cross-format locations).
#[allow(dead_code)]
pub fn locations_for(model: &ProjectModel, name: &str) -> Vec<Origin> {
    model.origins_for_name(name)
}

File: dart/lib/src/core/audit.dart
import '../config/config.dart';
import '../models/finding.dart';
import '../models/project_model.dart';
import 'exit_codes.dart';
import '../detectors/detector.dart';
import '../detectors/duplicates.dart';
import '../detectors/environment_diff.dart';
import '../detectors/missing.dart';
import '../detectors/public_prefix.dart';
import '../detectors/schema_validation.dart';
import '../detectors/type_mismatch.dart';
import '../detectors/typo.dart';
import '../detectors/undefined_source.dart';
import '../detectors/unused.dart';
import '../detectors/weak_secret.dart';

/// Detector ids in a well-known order. The order here is the order
/// findings are produced.
List<Detector> allDetectors() => [
      MissingDetector(),
      UnusedDetector(),
      UndefinedSourceDetector(),
      DuplicatesDetector(),
      EnvironmentDiffDetector(),
      TypeMismatchDetector(),
      PublicPrefixDetector(),
      WeakSecretDetector(),
      TypoDetector(),
      SchemaValidationDetector(),
    ];

class AuditOptions {
  final bool strict;
  final List<String> only;
  final Map<String, RuleSeverity> rules;

  AuditOptions({this.strict = false, this.only = const [], this.rules = const {}});
}

/// The audit engine is fully format-agnostic: it takes a [ProjectModel] and
/// returns an [AuditResult]. Detectors run in a stable order and their
/// findings are aggregated; the exit code is derived from severity (and
/// `strict`).
AuditResult runAudit(ProjectModel model, AuditOptions options) {
  final index = IndexedModel.buildIndex(model);

  final detectors =
      options.only.isNotEmpty ? allDetectors().where((d) => options.only.contains(d.id)).toList() : allDetectors();

  var findings = <Finding>[];
  for (final d in detectors) {
    findings.addAll(d.detect(index));
  }

  findings = applyIgnores(findings, model);
  findings = applyRuleSeverities(findings, options.rules);

  final errors = findings.where((f) => f.severity == Severity.error).length;
  final warnings = findings.where((f) => f.severity == Severity.warning).length;
  final infos = findings.where((f) => f.severity == Severity.info).length;
  final summary = AuditSummary(
    filesScanned: model.allFiles.length,
    variablesFound: _distinctVariableCount(model),
    errors: errors,
    warnings: warnings,
    infos: infos,
    total: findings.length,
  );

  final exitCode =
      ExitCodes.auditExitCode(ExitContext(findings: findings, strict: options.strict));

  return AuditResult(findings: findings, summary: summary, exitCode: exitCode);
}

/// Build a map of variable name → set of ignored rule ids from inline comments.
Map<String, Set<String>> _buildIgnoredRulesByName(ProjectModel model) {
  final ignored = <String, Set<String>>{};
  for (final file in model.envFiles) {
    if (file.environment == 'example') continue;
    for (final v in file.variables) {
      final rules = v.ignoreRules;
      if (rules == null || rules.isEmpty) continue;
      final set = ignored.putIfAbsent(v.name, () => <String>{});
      set.addAll(rules);
    }
  }
  return ignored;
}

/// Drop findings that are ignored inline in env files.
List<Finding> applyIgnores(List<Finding> findings, ProjectModel model) {
  final ignored = _buildIgnoredRulesByName(model);
  return findings.where((f) {
    final rules = ignored[f.variable];
    return !(rules != null && rules.contains(f.ruleId));
  }).toList();
}

/// Apply per-detector severity overrides from config. Disabled rules are dropped.
List<Finding> applyRuleSeverities(
    List<Finding> findings, Map<String, RuleSeverity> rules) {
  final result = <Finding>[];
  for (final f in findings) {
    final override = rules[f.ruleId];
    if (override == null) {
      result.add(f);
      continue;
    }
    switch (override) {
      case RuleSeverity.off:
        break;
      case RuleSeverity.error:
        final clone = f.clone()..severity = Severity.error;
        result.add(clone);
      case RuleSeverity.warning:
        final clone = f.clone()..severity = Severity.warning;
        result.add(clone);
    }
  }
  return result;
}

int _distinctVariableCount(ProjectModel model) {
  final names = <String>{};
  for (final file in model.allFiles) {
    for (final v in file.variables) {
      names.add(v.name);
    }
    for (final v in file.usages) {
      names.add(v.name);
    }
  }
  return names.length;
}
